src/OrchardCore.Modules/OrchardCore.Users/Controllers/AdminController.cs

```diff
@@ -293,11 +298,12 @@ private async Task SendActivationEmail(User user)
             if (sendActivationEmail != null && Convert.ToBoolean(sendActivationEmail))
             {
                 var token = await _userManager.GenerateEmailConfirmationTokenAsync(user);
-                var url = Url.Action("ActivateAccount", "Registration", new { Area = "OrchardCore.Users", email = user.Email, activationToken = token });
+                var path = Url.Action("ActivateAccount", "Registration", new { Area = "OrchardCore.Users", email = user.Email, activationToken = token });
                 var context = new AccountActivationContext(user)
                 {
-                    ActivationUrl = url
-                };
+                    // inject the tenant path
+                    ActivationUrl = PathExtensions.Combine(_shellSettings.Name, path)
```

Comment:
Normally the result of `Url.Action()` already contains the tenant prefix (and a virtual folder if there is one).

Reply:
Ok, and I should take the hostname from the request right?

Reply:
Might be better to take the hostname from the site settings base url property.
There's some security risks for spoofing from reading from the header (plus need to allow for x-forwarded if behind a load balancer, or cdn front etc)

Reply:
Is it always there?

Reply:
No it's not (which is annoying) and also see https://github.com/OrchardCMS/OrchardCore/issues/4836

For sitemaps I've used it, and am able to display in the ui a warning if it's empty
https://github.com/OrchardCMS/OrchardCore/pull/4450/files#diff-7f613c04a5336c9fa74dfe7fe9cf0acb

Up to you, of course :)

Reply:
> Ok, and I should take the hostname from the request right?

There are other `UrlHelper` / `LinkGenerator` helpers that provide uri strings and that grab for you the ambient values (as the host name), unless you explicitly pass them as parameters.

So it depends on the context and the ambient values that can be re-used or not. Hmm, here for the host name, as said by @deanmarcussen, better to pass a trusted value and the right one if behind something.

Note: But this is not what we do in `RegisterUserTask` using `_linkGenerator.GetUriByAction()` ;)

Reply:
actually I have a better look at the aspnetcore `.UseForwardedHeaders()` and it does provide good spoofing protection, but **we don't currently use that** in the reverse proxy module. It's a note for later. So we should probably do that ;)

There is a urlHelper extension for getting an absolute url `OrchardCore.Mvc.Core.Utilities.ToAbsoluteUrl` which we are making use of in other places, so maybe that is best. 

I probably continue to use it for sitemaps as it provides a useful way to prevent test/staging sites accidentally ending up in google, but that's a different issue.

Reply:
I think I've fixed it.
Next problem is that the Url in the email gets malformed to;
https://localhost:44300/OrchardCore.Users/Registration/ActivateAccount?email=info@localhost.com\u0026activationToken=CfDJ8Hs7Eh9x9Q5GkELorAMNqUsx2hZBa0siNwkI1hyHOcHTqyCAbs3f9QinNOKJeEHN5VrHDQnSTE4RCTaKe1oV7MzWcIV18SAkhb5RANC8GKg0nQfbrZP9sBbiaObGopIvFC2VVUEYSyjE4ip%2FX%2F0e2k7vzZPyWgO%2Fze2qROKvxsFMiSpzzPcP0w8pcCqt%2FPI1gQ%3D%3D 

The \u0026 is a problem, since it should be an &.
I've tried adding the pipe href option in generating the email, but to no avail.
![image](https://user-images.githubusercontent.com/1598506/71476344-dd79df80-27e4-11ea-837b-9b9150b449bf.png)

Reply:
@deanmarcussen we do use UseForwardedHeaders in reverseproxy module, btw.

Reply:
@MatthijsKrempel have you tried {{ | raw }}

yes, I know you use forwarded headers, so I thought you might be interested in the TODO features ;)

Reply:
It seems that all workflow serialization is done with the Javascript TextEncoder, I've made it optional so you can override it.
Also changed the Email task to render the body with the HtmlTextEncoder, that solves the issue.

Reply:
> Ok, and I should take the hostname from the request right?

I thought i had already put a comment but maybe i didn't post it ;) Yes there are some `UrlHelper /  LinkGenerator` helpers that provide full string uri and that use ambient values (as the host name) unless you explicitly pass them as parameters.

See in `RegisterUserTask` that uses the `LinkGenerator`.

For the host name, as @deanmarcussen said, better to provide a trusted value, and the right value if behind something. Hmm, but this is not what we are doing in `RegisterUserTask`.

Reply:
@jtkech will have a look!

Reply:
Used LinkBuilder seemed more appropriate